### ISSUE TITLE
windows.memory.aquisition

### ISSUE THREAD

**Issue by kyleEeeEEeeee** [Velocidex/velociraptor#3037]
Hello,

I'm trying to create an _offline_ collector that runs Kape and memory.  I can't get the memory part to work, output looks like below.  Can you please let me know what I'm doing wrong?

[INFO] 2023-10-18T14:16:59-04:00 Starting collection of Windows.Memory.Acquisition
[INFO] 2023-10-18T14:16:59-04:00 Adding global destructor for C:\Users\xxxxxxxxx\AppData\Local\Temp\tmp1960709620.exe
[ERROR] 2023-10-18T14:16:59-04:00 Symbol Filename not found. Current Scope is: [NULL], [_SessionId, $cache, $device_manager, config, $acl, $uploader, Artifact, $root], [Tool_WinPmem64_HASH, Tool_WinPmem64_FILENAME], [precondition_Windows_Memory_Acquisition_0], [Windows_Memory_Acquisition_0_0], [$Query], [$acl, $cache, config, $root, $uploader, $stack], [SleepDuration, ToolInfo, IsExecutable, ToolName], [SleepDuration], [IsExecutable], [RequiredTool], [matching_tools], [get_ext], [temp_binary]
[INFO] 2023-10-18T14:16:59-04:00 copy: Field filename Expecting a path arg type, not types.Null
[INFO] 2023-10-18T14:16:59-04:00 Adding global destructor for C:\Users\xxxxxxxxx\AppData\Local\Temp\tmp2861108791.raw
[INFO] 2023-10-18T14:16:59-04:00 shell: Running external command [Null C:\Users\xxxxxxx\AppData\Local\Temp\tmp2861108791.raw]
[INFO] 2023-10-18T14:16:59-04:00 shell: exec: "Null": executable file not found in %PATH%
[INFO] 2023-10-18T14:16:59-04:00 Collected 1 rows for Windows.Memory.Acquisition
[INFO] 2023-10-18T14:16:59-04:00 tempfile: removing tempfile C:\Users\xxxxxxxx\AppData\Local\Temp\tmp2861108791.raw
[INFO] 2023-10-18T14:16:59-04:00 tempfile: removing tempfile C:\Users\xxxxxxx\AppData\Local\Temp\tmp1960709620.exe
[DEBUG] 2023-10-18T14:16:59-04:00 Query Stats: {"RowsScanned":5,"PluginsCalled":5,"FunctionsCalled":2,"ProtocolSearch":4,"ScopeCopy":17}
[INFO] 2023-10-18T14:16:59-04:00 Container hash f4db2cf02da577fc8ae482c6a51475ef6c86767b3f82d7c4946e3a40616c264e
[DEBUG] 2023-10-18T14:16:59-04:00 Query Stats: {"RowsScanned":33,"PluginsCalled":8,"FunctionsCalled":85,"ProtocolSearch":116,"ScopeCopy":79}
[
 {
  "Container": "C:/Users/xxxxxxxxxx/Desktop/server-F.CKO20SB73DA3U/uploads/Collection-SC10PSECPEN01_xxxxxxxx-2023-10-18T14_16_59-04_00.zip"
 }
][INFO] 2023-10-18T14:16:59-04:00 Collection completed in 0 Seconds
[DEBUG] 2023-10-18T14:16:59-04:00 Query Stats: {"RowsScanned":1,"PluginsCalled":1,"FunctionsCalled":0,"ProtocolSearch":0,"ScopeCopy":3}
[INFO] 2023-10-18T14:16:59-04:00 Exiting notification service for Root Org!
PS C:\Users\xxxxxxxxxx\Desktop\server-F.CKO20SB73DA3U\uploads>

**Comment by scudette** [comment]
How are you generating the collector? Using the GUI? What version are you running?

**Comment by kyleEeeEEeeee** [comment]
Hello, yes I am using the GUI and version "version":"0.7.0-2"

**Comment by kyleEeeEEeeee** [comment]
does the Endpoint Filename have to match the name of the upstream dowloaded tool?

**Comment by scudette** [comment]
It looks like the collector that was created is missing the zip file that contains the winpmem binary
this is what it is supposed to looks like when just adding the acquisition artifact by itself

```
C:\Users\Administrator\Downloads>Collector_velociraptor-v0.7.0-3-windows-amd64.exe
[INFO] 2023-10-18T06:13:08-07:00 Autoexec with parameters: [artifacts collect Collector --logfile Collector_velociraptor-v0.7.0-3-windows-amd64.exe.log -v --require_admin]
[INFO] 2023-10-18T06:13:08-07:00  _    __     __           _                  __
[INFO] 2023-10-18T06:13:08-07:00 | |  / /__  / /___  _____(_)________ _____  / /_____  _____
[INFO] 2023-10-18T06:13:08-07:00 | | / / _ \/ / __ \/ ___/ / ___/ __ `/ __ \/ __/ __ \/ ___/
[INFO] 2023-10-18T06:13:08-07:00 | |/ /  __/ / /_/ / /__/ / /  / /_/ / /_/ / /_/ /_/ / /
[INFO] 2023-10-18T06:13:08-07:00 |___/\___/_/\____/\___/_/_/   \__,_/ .___/\__/\____/_/
[INFO] 2023-10-18T06:13:08-07:00                                   /_/
[INFO] 2023-10-18T06:13:08-07:00 Digging deeper!                  https://www.velocidex.com
[INFO] 2023-10-18T06:13:08-07:00 This is Velociraptor 0.7.0-3 built on 2023-10-01T13:24:20Z (4c0007b2)
[INFO] 2023-10-18T06:13:08-07:00 Loaded embedded config
[INFO] 2023-10-18T06:13:08-07:00 Starting Org Manager service.
[INFO] 2023-10-18T06:13:08-07:00 Starting services for Root Org
[INFO] 2023-10-18T06:13:08-07:00 Starting Journal service for Root Org.
[INFO] 2023-10-18T06:13:08-07:00 Starting the notification service for Root Org.
[INFO] 2023-10-18T06:13:08-07:00 Installing Dummy inventory_service. Will download tools to temp directory.
[INFO] 2023-10-18T06:13:08-07:00 Starting repository manager for Root Org
[INFO] 2023-10-18T06:13:08-07:00 Loaded 375 built in artifacts in 481.1478ms
[INFO] 2023-10-18T06:13:08-07:00 Adding tool WinPmem64
[INFO] 2023-10-18T06:13:08-07:00 Output Prefix : C:/Users/Administrator/Downloads
[INFO] 2023-10-18T06:13:08-07:00 Will collect package C:/Users/Administrator/Downloads/Collection-WIN-NBFP7TTP911_lan-2023-10-18T06_13_08-07_00.zip
[INFO] 2023-10-18T06:13:08-07:00 Starting collection of Collector
[INFO] 2023-10-18T06:13:08-07:00 Setting compression level to 5
[INFO] 2023-10-18T06:13:08-07:00 Will create container at C:/Users/Administrator/Downloads/Collection-WIN-NBFP7TTP911_lan-2023-10-18T06_13_08-07_00.zip
[INFO] 2023-10-18T06:13:08-07:00 Starting collection of Windows.Memory.Acquisition
[INFO] 2023-10-18T06:13:08-07:00 Adding global destructor for C:\Users\ADMINI~1\AppData\Local\Temp\2\tmp157201299.exe
[INFO] 2023-10-18T06:13:08-07:00 copy: Copying file from /uploads/winpmem_mini_x64_rc2.exe into C:\Users\ADMINI~1\AppData\Local\Temp\2\tmp157201299.exe
[INFO] 2023-10-18T06:13:08-07:00 Adding global destructor for C:\Users\ADMINI~1\AppData\Local\Temp\2\tmp583337716.raw
[INFO] 2023-10-18T06:13:08-07:00 shell: Running external command [C:\Users\ADMINI~1\AppData\Local\Temp\2\tmp157201299.exe C:\Users\ADMINI~1\AppData\Local\Temp\2\tmp583337716.raw]
[INFO] 2023-10-18T06:13:09-07:00 Collecting file C:\Users\ADMINI~1\AppData\Local\Temp\2\tmp583337716.raw into /uploads/auto/PhysicalMemory.raw (0 bytes)
[INFO] 2023-10-18T06:13:09-07:00 Collected 6 rows for Windows.Memory.Acquisition
[INFO] 2023-10-18T06:13:09-07:00 tempfile: removing tempfile C:\Users\ADMINI~1\AppData\Local\Temp\2\tmp583337716.raw
[INFO] 2023-10-18T06:13:09-07:00 tempfile: removing tempfile C:\Users\ADMINI~1\AppData\Local\Temp\2\tmp157201299.exe
[DEBUG] 2023-10-18T06:13:09-07:00 Query Stats: {"RowsScanned":20,"PluginsCalled":5,"FunctionsCalled":8,"ProtocolSearch":4,"ScopeCopy":47}
[INFO] 2023-10-18T06:13:09-07:00 Container hash 1b1a19ab2fe75b933c83df95bdab2e25a99345029e9d159fe8eccfbb84772e64
[DEBUG] 2023-10-18T06:13:09-07:00 Query Stats: {"RowsScanned":33,"PluginsCalled":8,"FunctionsCalled":85,"ProtocolSearch":116,"ScopeCopy":79}
[
 {
  "Container": "C:/Users/Administrator/Downloads/Collection-WIN-NBFP7TTP911_lan-2023-10-18T06_13_08-07_00.zip"
 }
][INFO] 2023-10-18T06:13:09-07:00 Collection completed in 1 Seconds
[DEBUG] 2023-10-18T06:13:09-07:00 Query Stats: {"RowsScanned":1,"PluginsCalled":1,"FunctionsCalled":0,"ProtocolSearch":0,"ScopeCopy":3}
[INFO] 2023-10-18T06:13:09-07:00 Exiting notification service for Root Org!
```

It still fails because it looks like winpmem is not working on this system but it cn extract it with no issue.

you can check the zip file that should be at the end of the collector in windows like this

```
C:\Users\Administrator\Downloads>tar -tf Collector_velociraptor-v0.7.0-3-windows-amd64.exe
uploads/winpmem_mini_x64_rc2.exe
uploads/inventory.csv
```

**Comment by kyleEeeEEeeee** [comment]
Hi! Sorry about the delay.  I spent a while troubleshooting after we talked.  I can get it to work now, but it seems to be hit or miss and I can't put my finger on why.  Seems to be less reliable on VMs, but works on some vs others even with the same build.

**Comment by kyleEeeEEeeee** [comment]
FWIW, when I built the executable on a non-vm Windows GUI, it started to work, then had the issue after that with it sometimes working and sometimes not.

**Comment by scudette** [comment]
Are you saying that the built binary that the GUI produced was missing the zip file?

Does it work with the new generic collector mode? That does not repack the binary itself only a small external file.

Can you maybe share one of the badly built binaries so we can see what is corrupted in the zip file creation? Maybe it's some kind of bug in building the zip file and appending it to the end of the file.

**Comment by kyleEeeEEeeee** [comment]
Will try the generic collector today, I had not tried that yet! And yes I think I should be able to share the binary that did not work.  Can I send it to you directly?

**Comment by scudette** [comment]
Sure - feel free to dm on discord or send to support@velocidex.com

**Comment by scudette** [comment]
This was a bug in the 0.7.0 release which was fixed in 0.7.0-4
